### PR TITLE
Project Finished!

### PR DESCRIPTION
**Pull request by HarryjLeigh**
This was an enjoyable project to work on! Looking forward to your code review - Don't hold back :)

### CODE REVIEW

File: HarryJLeigh.FlashCards/FlashCards/Controllers/FlashcardController.cs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+namespace FlashCards.Controllers;
+
+public class FlashcardController
+{
+    private DatabaseService DbService { get; }
+    
+    public FlashcardController(DatabaseService dbService) => DbService = dbService;
+    
+    public List<FlashCardDto> GetAllFlashcards(int stack_id) => ExecuteQuery("Select flashCardId, front, back FROM Flashcards WHERE stack_id = @Stack_id", new {Stack_id = stack_id });
```
Comment (chrisjamiecarter) [inline HarryJLeigh.FlashCards/FlashCards/Controllers/FlashcardController.cs:13]: 🟠 **Code Readability**

💡 If you are going to smash all this together, at least make it readable... Blank lines, line length limit 😄 please 😆

File: HarryJLeigh.FlashCards/FlashCards/Controllers/FlashcardController.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+using Dapper;
+using FlashCards.Data;
+
+namespace FlashCards.Controllers;
+
+public class FlashcardController
+{
+    private DatabaseService DbService { get; }
+    
+    public FlashcardController(DatabaseService dbService) => DbService = dbService;
```
Comment (chrisjamiecarter) [inline HarryJLeigh.FlashCards/FlashCards/Controllers/FlashcardController.cs:11]: 🟠 **Code Style**

💡 If you are going to style your constructor like this, you might as well make it a [primary constructor](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/primary-constructors).

File: HarryJLeigh.FlashCards/FlashCards/Controllers/FlashcardController.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
+using System.Data.SqlClient;
```
Comment (chrisjamiecarter) [inline HarryJLeigh.FlashCards/FlashCards/Controllers/FlashcardController.cs:1]: 🟠 **Deprecated**

💡 You should be using `Microsoft.Data.SqlClient`

image

File: HarryJLeigh.FlashCards/FlashCards/README.md
Comment (chrisjamiecarter) [inline HarryJLeigh.FlashCards/FlashCards/README.md:1]: 🟢 **README File**

⭐ Fantastic Readme file, well done!

File: HarryJLeigh.FlashCards/FlashCards/README.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+## PROJECT OVERVIEW
+This is a console-based CRUD application for creating, managing, and studying flashcards, built with **C#**, **Dapper**, **MSSQL**, and **Spectre.Console**. The application allows users to create flashcards, organise them into stacks, and study using an intuitive console-based UI.
+
+The application is designed to improve learning by providing features like:
+- Flashcard creation and organisation.
+- Interactive study sessions with progress tracking.
+- Comprehensive reporting.
+
+**Note**:
```
Comment (chrisjamiecarter) [inline HarryJLeigh.FlashCards/FlashCards/README.md:11]: 🟠 **Missing Step**

💡 Remember you need to instruct the user to create the database too if your app doesn't do it on startup.

**Review by chrisjamiecarter — CHANGES_REQUESTED** [review]
Hey @HarryjLeigh 👋,

Excellent work on your Flashcards project submission 🎉!
I have performed a peer review and have added some praise and suggestions as bespoke comments.
Also, see below for some general feedback.

***

🟠 **Solution Structure**
Close! But as per the academy's [Code Reviews](https://thecsharpacademy.com/article/52/code-reviews) page. You project submission should be as per the below.
```
.
└── [ProjectName].[GitHubUserName]/
    ├── [SolutionDirectory1]/
    ├── [SolutionDirectory2]/
    └── [ProjectName].sln
```
This will make it easier for code-reviewers to pull down and review your code in the future.

So: `Flashcards.HarryjLeigh`

🟠 **User Experience**
Give an option to exit, once a user selects an option they are committed.

image

Stuck if no stacks/flashcards exist

image


🔴 **Stack Name Validation**
If you update a stack name, you can side step the name validation.

image

***

Please revisit to try understand and fix any issues 🔧 , as these relate to the project requirements they block my approval.

Feel free to reach out if you have any questions or if you'd like to collaborate further on any of these points 🆘 .

I will go ahead and wait for you to make your changes, commit and push (which will then notify me) 😊 !

Best regards,
@chrisjamiecarter 👍

File: HarryJLeigh.FlashCards/FlashCards/Controllers/FlashcardController.cs (hunk 1 above)
Comment (HarryjLeigh) [reply HarryJLeigh.FlashCards/FlashCards/Controllers/FlashcardController.cs]: Looking back at this, I realise the long line is quite hard to read. I've improved the readability by breaking it into multiple lines for better clarity. 😊

File: HarryJLeigh.FlashCards/FlashCards/Controllers/FlashcardController.cs (hunk 2 above)
Comment (HarryjLeigh) [reply HarryJLeigh.FlashCards/FlashCards/Controllers/FlashcardController.cs]: Newly learned about primary constructors and they're great! Added to the  flashcard, report and stack controllers

File: HarryJLeigh.FlashCards/FlashCards/Controllers/FlashcardController.cs (hunk 3 above)
Comment (HarryjLeigh) [reply HarryJLeigh.FlashCards/FlashCards/Controllers/FlashcardController.cs]: Removed and added Microsoft.Data.SqlClient

File: HarryJLeigh.FlashCards/FlashCards/README.md (hunk 4 above)
Comment (HarryjLeigh) [reply HarryJLeigh.FlashCards/FlashCards/README.md]: Updated ReadMe with database creation.

**Comment by HarryjLeigh** [comment]
I tried to fix the solution structure, but it doesn’t seem to have taken effect 😅. However, the rest of the feedback has been addressed! 

Thank you for your feedback! It has helped me recognize aspects I wouldn’t have considered including in the project. 😊

**Review by chrisjamiecarter — APPROVED** [review]
Hey @Harryjl23 👋,

Great job getting the updates pushed through 🦾 

You now meet all of the project requirements ⭐ 


***

Feel free to reach out if you have any questions or if you'd like to collaborate further on any of these points 🆘 .

I will go ahead and mark as approved, keep up the excellent work on the next projects! 😊

Best regards,
@chrisjamiecarter 👍

**Pull request closed without merging**